### ISSUE TITLE
array.reset causes array.size() to be set to 8

### ISSUE THREAD

**Issue by dbadb** [ccrma/chuck#166]
`array.reset` causes `array.size()` to return 8. 

Seems to me an array .reset() should result in a 0-sized array but others may disagree?  Docs say:

```
Reset array to original state; clears the array and sets capacity to 8.
```

nb: no problem for me on the capacity, just the size.

**Comment by gewang** [comment]
Thank you for reporting this!  This has been addressed and will be included in 1.4.0.2.  Also added array.capacity (read/write).
FYI in the future, .cap() will be deprecated (it actually behaves identically to .size(), which is confusing!) — i.e., should use either .size or .capacity, depending on the context; .cap() should be avoided moving forward.

**Comment by dbadb** [comment]
@gewang - on a related note.  I'm getting some crashers on windows debug builds when I perform lots of array resizes.  My sense is that the range checks you perform via capacity should actually be based on size, not capacity.

```c++
 if( i < 0 || i >= m_vector.capacity())
 ...
else
   return m_vector[i];  // if capacity > size(), this can lead to bad access
```
From a windows + stl + debug point of view, capacity is a memory-allocation optimization but it's an error (at least when debugging is enabled) to access a std::vector outside it's current size.

I've gone into chuck_oo.cpp and replaced the bounds checks with size() rather than capacity() and my windows crashers go away.

Thoughts?  One might argue that this is just windows pedantry. I see that you are legit clearing array values on down-size resets (which seems conceptually fine).  On the other hand, user-code should probably not be able to access beyond size().
